qelem/lib/infrastructure/question/question_repository.dart

```diff
@@ -0,0 +1,161 @@
+    } on QHttpException catch (e) {
+      return Either(error: Error(e.message));
+    }
+  }
+
+  // get question by id
+  Future<Either<Question>> getQuestionById(int id) async {
+    try {
+      QuestionDto questionDto = await questionApi.getQuestionById(id);
+      return Either(val: questionDto.toQuestion());
+    } on QHttpException catch (e) {
+      return Either(error: Error(e.message));
+    }
+  }
+
+  // create question
+  Future<Either<Question>> createQuestion(Question question) async {
+    try {
+      QuestionDto questionDto = await questionApi.createQuestion(
+          QuestionFormDto(topic: question.topic, content: question.content));
+      return Either(val: questionDto.toQuestion());
+    } on QHttpException catch (e) {
+      return Either(error: Error(e.message));
+    }
+  }
+
+  // delete question
+  Future<Either<void>> deleteQuestion(int id) async {
+    try {
+      await questionApi.deleteQuestion(id);
+      return Either(val: null);
+    } on QHttpException catch (e) {
+      return Either(error: Error(e.message));
+    }
+  }
+
+  // update question
+  Future<Either<Question>> updateQuestion(Question question) async {
+    try {
+      QuestionDto questionDto = await questionApi.updateQuestion(
+          QuestionFormDto(topic: question.topic, content: question.content),
+          question.id);
+      return Either(val: questionDto.toQuestion());
+    } on QHttpException catch (e) {
+      return Either(error: Error(e.message));
+    }
+  }
+
+  // upVote question
+  Future<Either<Question>> upvoteQuestion(Question question) async {
+    int prevVote;
+    if (question.userVote == Vote.upVote) {
+      prevVote = 1;
+    } else if (question.userVote == Vote.downVote) {
+      prevVote = -1;
+    } else {
+      prevVote = 0;
+    }
+
+    if (question.userVote == Vote.upVote) {
+      try {
+        await questionApi.unvoteQuestion(question.id);
+
+        return Either(
+            val: Question(
+                id: question.id,
+                topic: question.topic,
+                content: question.content,
+                upVotes: question.upVotes - prevVote, // prevVote = 1
+                userVote: Vote.none,
+                downVotes: question.downVotes,
+                author: question.author));
+      } on QHttpException catch (e) {
+        return Either(error: Error(e.message));
+      }
+    } else {
+      try {
+        await questionApi.upVoteQuestion(question.id);
+
+        return Either(
+            val: Question(
+                id: question.id,
+                topic: question.topic,
+                content: question.content,
+                upVotes: question.upVotes + 1,
+                userVote: Vote.upVote,
+                downVotes: question.downVotes + prevVote, // prevVote = -1 or 0
+                author: question.author));
+      } on QHttpException catch (e) {
+        return Either(error: Error(e.message));
+      }
+    }
+  }
+
```

Comment:
What i was suggesting is that what if we send a request to the server( like questionApi.getQuestionById) after up-voting  and return a Question model rather than creating and tweaking a new question model in the repository.

Reply:
@Biruk-Tassew I was trying to avoid the bandwidth. Both way works, but I thought it would be better to avoid data cost. What do you think?

Reply:
@hello-world-was-taken  Yeah...i'll edit mine